# Проект SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Техническое задание

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
- Опредить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### 1 Доступ к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''

pd.io.sql.read_sql(sql=text(query), con = engine)

book_id  author_id                                              title   
0        1        546                                       'Salem's Lot  \
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Доступ к данным получен, можно продолжать исследование.

### 2 Изучение данных

In [2]:
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']

In [3]:
for name in tables:
    query = f'''SELECT * FROM {name} LIMIT 5'''
    print('\n')
    print(' '*40, f'Первые 5 строк таблицы - {name}')
    display(pd.io.sql.read_sql(sql=text(query), con = engine))



                                         Первые 5 строк таблицы - books


book_id  author_id                                              title   
0        1        546                                       'Salem's Lot  \
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268



                                         Первые 5 строк таблицы - authors


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd



                                         Первые 5 строк таблицы - publishers


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company



                                         Первые 5 строк таблицы - ratings


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2



                                         Первые 5 строк таблицы - reviews


review_id  book_id       username   
0          1        1   brandtandrea  \
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

##### Выведенные таблицы совпадают с описанием данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

##### Функция для быстрого набора

In [4]:
print(tables)

['books', 'authors', 'publishers', 'ratings', 'reviews']


In [5]:
def request(query):
    display(pd.io.sql.read_sql(sql=text(query), con = engine))

##### Количество книг, вышедших после 1 января 2000 года

In [6]:
query = '''
SELECT COUNT(book_id)
FROM books 
WHERE publication_date > '2000-01-01'
'''

In [7]:
request(query)

count
0    819

Начиная со 1 января 2000 года всего вышло 819 книг.

##### Количество обзоров и среднюю оценку каждой книги

In [8]:
query = '''
WITH temp_1 AS (
SELECT book_id,
       ROUND(AVG(rating), 2) AS avg_rating
FROM ratings 
GROUP BY book_id),

temp_2 AS (
SELECT book_id,
       COUNT(review_id) AS reviews_cnt
FROM reviews 
GROUP BY book_id)

SELECT t1.book_id,
       t1.avg_rating,
       t2.reviews_cnt
FROM temp_1 AS t1
LEFT JOIN temp_2 AS t2 ON t1.book_id=t2.book_id
'''

In [9]:
request(query)

book_id  avg_rating  reviews_cnt
0          1        3.67          2.0
1          2        2.50          1.0
2          3        4.67          3.0
3          4        4.50          2.0
4          5        4.00          4.0
..       ...         ...          ...
995      996        3.67          3.0
996      997        3.40          3.0
997      998        3.20          4.0
998      999        4.50          2.0
999     1000        3.83          4.0

[1000 rows x 3 columns]

Выведены для id каждой книги: 
- количество обзоров
- средняя оценка

##### Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [10]:
query = '''
SELECT publisher
FROM publishers
WHERE publisher_id = (SELECT publisher_id
                             FROM books 
                             WHERE num_pages > 50
                             GROUP BY publisher_id
                             ORDER BY COUNT(book_id) DESC
                             LIMIT 1)
   
       
'''

In [11]:
request(query)

publisher
0  Penguin Books

Издательство, которое выпустило наибольшее число книг толще 50 страниц - **Penguin Books**

##### Автор с самой высокой средней оценкой книг (учитываются только книги с 50 и более оценками)

In [12]:
query = '''
SELECT author,
       ROUND(AVG(rating), 2) as avg_rating
FROM authors AS a
INNER JOIN books AS b ON b.author_id=a.author_id
INNER JOIN ratings AS r ON r.book_id=b.book_id
WHERE b.book_id IN (SELECT  book_id
                            FROM ratings
                            GROUP BY book_id
                            HAVING COUNT(rating) > 50)
GROUP BY author
ORDER BY AVG(rating) DESC
LIMIT 1
'''

In [13]:
request(query)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

**J.K. Rowling/Mary GrandPré** - автор с самой высокой средней оценкой - 4.29.

##### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [14]:
query = '''
WITH temp AS 
   (SELECT username,
           COUNT(review_id) AS review_cnt
    FROM reviews
    WHERE username IN (SELECT username
                              FROM ratings
                              GROUP BY username
                              HAVING COUNT(rating) > 50)
    GROUP BY username)
    
SELECT ROUND(AVG(review_cnt), 2)
FROM temp
'''

In [15]:
request(query)

round
0  24.33

24.33 - среднее количество обзоров от пользователей, которые поставили больше 50 оценок.